### PR TITLE
Add support for dependencies when publishing to ACR server

### LINKED ISSUE #1516: Publish support/testing for modules with dependencies in ACR
### Summary of the new feature / enhancement

Publish support/testing for modules with dependencies in ACR

### Proposed technical implementation details (optional)

_No response_

### PR DESCRIPTION
**Pull request by alerickson**
# PR Summary
This PR implements dependency support for publishing modules to ACR/container repositories.

This implementation attaches an artifact ("dependency") to the module being published that contains a file called dependency.json which lists all the first level dependencies for the parent package (retrieved from the module manifest of the module being published).  The "dependency" artifact also has its own manifest (pictured below).

image

image

  
 

## PR Context
Partially resolves #1516 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/ACRServerAPICalls.cs [hunk 1]
```diff
@@ -1177,67 +1231,220 @@ internal bool PushNupkgACR(string psd1OrPs1File, string outputNupkgDir, string p
...
+                }
+
+                _cmdletPassedIn.WriteVerbose("Create the manifest");
+
+                // Upload Manifest
+                var depManifestResponse = UploadDependencyManifest(pkgNameLower, $"sha256:{artifactDigest}", depConfigFilePath, true, acrAccessToken).Result;
+                bool depManifestCreated = depManifestResponse.IsSuccessStatusCode;
+                if (!depManifestCreated)
+                {
+                    // TODO: error out
```
Comment (adityapatwardhan) [inline src/code/ACRServerAPICalls.cs]: Add the error out

File: src/code/ACRServerAPICalls.cs [hunk 2]
```diff
@@ -1177,67 +1231,220 @@ internal bool PushNupkgACR(string psd1OrPs1File, string outputNupkgDir, string p
...
+                    return false;
+                }
+
+                _cmdletPassedIn.WriteVerbose("End of dependency processing");
+            }
+            catch
+            {
+                // TODO: throw
```
Comment (adityapatwardhan) [inline src/code/ACRServerAPICalls.cs]: Terminating error?

File: src/code/ACRServerAPICalls.cs [hunk 3]
```diff
@@ -1177,67 +1231,220 @@ internal bool PushNupkgACR(string psd1OrPs1File, string outputNupkgDir, string p
...
             }
+            long parentSize = (long)manifestResponse.RequestMessage.Content.Headers.ContentLength;
+
+            // Create manifest for dependencies
+            var depJsonStr = depJsonContent.Replace("\r", String.Empty).Replace("\n", String.Empty);
+            string depFileContent = CreateDependencyManifestContent(emptyConfigArtifactDigest, 0, depDigest, depFileSize, depFileName, depJsonStr, parentDigest, parentSize);
+            File.WriteAllText(depConfigFilePath, depFileContent);
+
+            var depManifestDigestCreated = CreateDigest(depConfigFilePath, out artifactDigest, out ErrorRecord artifactDigesterror);
+            // TODO: add err handling
```
Comment (adityapatwardhan) [inline src/code/ACRServerAPICalls.cs]: Add error handling

File: src/code/ACRServerAPICalls.cs [hunk 4]
```diff
@@ -1291,11 +1500,105 @@ internal bool PushNupkgACR(string psd1OrPs1File, string outputNupkgDir, string p
...
+
+            return stringWriter.ToString();
+        }
+
         // ACR method
         private bool CreateDigest(string fileName, out string digest, out ErrorRecord error)
         {
             FileInfo fileInfo = new FileInfo(fileName);
             SHA256 mySHA256 = SHA256.Create();
+
```
Comment (adityapatwardhan) [inline src/code/ACRServerAPICalls.cs:1621]:
```suggestion
```

File: src/code/ACRServerAPICalls.cs [hunk 5]
```diff
@@ -1312,7 +1615,55 @@ private bool CreateDigest(string fileName, out string digest, out ErrorRecord er
...
+                {
+                    var AuthorizationError = new ErrorRecord(ex, $"UnauthorizedAccessException for .nupkg file: {ex.Message}", ErrorCategory.PermissionDenied, null);
+                    error = AuthorizationError;
+                }
+            }
+            if (error != null)
+            {
+                return false;
+            }
+
```
Comment (adityapatwardhan) [inline src/code/ACRServerAPICalls.cs:1656]:
```suggestion
```

File: src/code/ACRServerAPICalls.cs [hunk 6]
```diff
@@ -1330,6 +1681,10 @@ private bool CreateDigest(string fileName, out string digest, out ErrorRecord er
             {
                 return false;
             }
+
+
+            Console.WriteLine($"digest: {digest}");
```
Comment (adityapatwardhan) [inline src/code/ACRServerAPICalls.cs]: Remove?

File: src/code/PSGetException.cs [hunk 7]
```diff
@@ -68,4 +68,12 @@ public LocalResourceNotFoundException(string message, Exception innerException =
         {
         }
     }
+
+    public class ProcessDependencyException : Exception
```
Comment (adityapatwardhan) [inline src/code/PSGetException.cs:72]: unsed class?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Comment by adityapatwardhan** [comment]
@alerickson - also need to add tests for this

File: src/code/PSGetException.cs (hunk 7 above)
Comment (alerickson) [reply src/code/PSGetException.cs:72]: Updated where this should have gone.

File: src/code/ACRServerAPICalls.cs [hunk 8]
```diff
@@ -1177,67 +1230,240 @@ internal bool PushNupkgACR(string psd1OrPs1File, string outputNupkgDir, string p
...
+        {
             _cmdletPassedIn.WriteVerbose("Create the config file");
-            string configFileName = "config.json";
-            var configFilePath = System.IO.Path.Combine(outputNupkgDir, configFileName);
             while (File.Exists(configFilePath))
             {
                 configFilePath = Guid.NewGuid().ToString() + ".json";
             }
-            using (FileStream configStream = new FileStream(configFilePath, FileMode.Create)){ }
+            File.Create(configFilePath).Dispose();
```
Comment (anamnavi) [inline src/code/ACRServerAPICalls.cs]: What happens if the file can't be created? The FileStream may be null then and `.Dispose()` would be called on that. Perhaps the file should be created in a try catch and then if the FileStream isn't null then dispose.

**Review by anamnavi — CHANGES_REQUESTED** [review]

File: src/code/ACRServerAPICalls.cs (hunk 8 above)
Comment (alerickson) [reply src/code/ACRServerAPICalls.cs]: Good catch, just added some exception handling in a new method

**Review by anamnavi — APPROVED** [review]
LGTM

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**